File: server/saki/messages.py
from enum import Enum
from dataclasses import dataclass
from typing import List, Tuple
import uuid
import json
from json import JSONEncoder

class DataClassJSONEncoder(JSONEncoder):
    def default(self, o):
        return o.__dict__


class SAKIException(Exception):
    pass

class SAKIDecodingException(SAKIException):
    pass

class SAKIRemoteException(SAKIException):
    pass

class SAKIMessageType(Enum):
    ERROR                   = "error"
    IMPORT_KEY              = "importKey"
    IMPORT_KEY_RESULT       = "importKeyResult"
    GENERATE_KEY            = "generateKey"
    GENERATE_KEY_RESULT     = "generateKeyResult"
    GET_ENTRY               = "getEntry"
    GET_ENTRY_RESULT        = "getEntryResult"
    LIST_KEYS               = "listKeys"
    LIST_KEYS_RESULT        = "listKeysResult"
    ENCRYPT_DATA            = "encryptData"
    ENCRYPT_DATA_RESULT     = "encryptDataResult"
    DECRYPT_DATA            = "decryptData"
    DECRYPT_DATA_RESULT     = "decryptDataResult"
    SIGN_DATA               = "signData"
    SIGN_DATA_RESULT        = "signDataResult"
    VERIFY_SIGNATURE        = "verifySignature"
    VERIFY_SIGNATURE_RESULT = "verifySignatureResult"
    

@dataclass
class SAKIMessage(object):
    id: str
    messageType: str
    data: dict

    def encode(self):
        return json.dumps(self.__dict__, cls=DataClassJSONEncoder)
    

    @staticmethod
    def decode(message: str):
        return SAKISubMessage.decode(message)


class SAKISubMessage(object):
    TYPE = None

    def build_message(self, id: str=None):
        return SAKIMessage(id or str(uuid.uuid4()), self.TYPE.value, json.dumps(self, cls=DataClassJSONEncoder))


    @staticmethod
    def decode(message: str):
        saki_msg = json.loads(message)

        for subclass in SAKISubMessage.__subclasses__():
            if subclass.TYPE.value == saki_msg['messageType']:
                return subclass(**json.loads(saki_msg['data']))
        
        raise SAKIDecodingException(f"Unknown request type: {saki_msg['messageType']}")



@dataclass
class SAKIError(SAKISubMessage):
    TYPE = SAKIMessageType.ERROR
    message: str


@dataclass
class AlgorithmSpecMessage(SAKISubMessage):
    TYPE = SAKIMessageType.GENERATE_KEY
    algorithm: str
    keyAlias: str
    purposes: int
    keySize: int
    digests: List[str]
    encryptionPaddings: List[str]


@dataclass
class ImportKeyMessage(SAKISubMessage):
    TYPE = SAKIMessageType.IMPORT_KEY
    wrappedKeyAlias: str
    wrappedKey: str
    transformation: str
    spec: AlgorithmSpecMessage


@dataclass
class ImportKeyResult(SAKISubMessage):
    TYPE = SAKIMessageType.IMPORT_KEY_RESULT
    success: bool


@dataclass
class GenerateKeyResult(SAKISubMessage):
    TYPE = SAKIMessageType.GENERATE_KEY_RESULT
    key: str


@dataclass
class GetEntryMessage(SAKISubMessage):
    TYPE = SAKIMessageType.GET_ENTRY
    keyAlias: str


@dataclass
class GetEntryResult(SAKISubMessage):
    TYPE = SAKIMessageType.GET_ENTRY_RESULT
    key: str
    attributes: List[Tuple[str, str]]


@dataclass
class ListKeysMessage(SAKISubMessage):
    TYPE = SAKIMessageType.LIST_KEYS


@dataclass
class ListKeysResult(SAKISubMessage):
    TYPE = SAKIMessageType.LIST_KEYS_RESULT
    keyAliases: List[str]


@dataclass
class EncryptDataMessage(SAKISubMessage):
    TYPE = SAKIMessageType.ENCRYPT_DATA
    keyAlias: str
    transformation: str
    data: str


@dataclass
class EncryptDataResult(SAKISubMessage):
    TYPE = SAKIMessageType.ENCRYPT_DATA_RESULT
    data: str


@dataclass
class DecryptDataMessage(SAKISubMessage):
    TYPE = SAKIMessageType.DECRYPT_DATA
    keyAlias: str
    transformation: str
    data: str


@dataclass
class DecryptDataResult(SAKISubMessage):
    TYPE = SAKIMessageType.DECRYPT_DATA_RESULT
    data: str


@dataclass
class SignDataMessage(SAKISubMessage):
    TYPE = SAKIMessageType.SIGN_DATA
    keyAlias: str
    transformation: str
    data: str


@dataclass
class SignDataResultMessage(SAKISubMessage):
    TYPE = SAKIMessageType.SIGN_DATA_RESULT
    signature: str


@dataclass
class VerifySignatureMessage(SAKISubMessage):
    TYPE = SAKIMessageType.VERIFY_SIGNATURE
    keyAlias: str
    transformation: str
    data: str
    signature: str


@dataclass
class VerifySignatureResult(SAKISubMessage):
    TYPE = SAKIMessageType.VERIFY_SIGNATURE_RESULT
    verified: bool

File: client/app/src/main/java/com/bobani/saki/Messages.kt
package com.bobani.saki

import kotlinx.serialization.*

class SAKIMessageType {
    companion object {
        const val ERROR     = "error"
        const val IMPORT_KEY = "importKey"
        const val IMPORT_KEY_RESULT = "importKeyResult"
        const val GENERATE_KEY = "generateKey"
        const val GENERATE_KEY_RESULT = "generateKeyResult"
        const val GET_ENTRY = "getEntry"
        const val GET_ENTRY_RESULT = "getEntryResult"
        const val LIST_KEYS = "listKeys"
        const val LIST_KEYS_RESULT = "listKeysResult"
        const val ENCRYPT_DATA = "encryptData"
        const val ENCRYPT_DATA_RESULT = "encryptDataResult"
        const val DECRYPT_DATA = "decryptData"
        const val DECRYPT_DATA_RESULT = "decryptDataResult"
        const val SIGN_DATA = "signData"
        const val SIGN_DATA_RESULT = "signDataResult"
        const val VERIFY_SIGNATURE = "verifySignature"
        const val VERIFY_SIGNATURE_RESULT = "verifySignatureResult"
    }
}


@Serializable
data class SAKIMessage(val id: String, val messageType: String, val data: String)

@Serializable
data class SAKIError(val message: String)

@Serializable
data class AlgorithmSpecMessage(val algorithm: String, val keyAlias: String, val purposes: Int, val keySize: Int, val digests: List<String>, val encryptionPaddings: List<String>)

@Serializable
data class GenerateKeyResult(val key: String)

@Serializable
data class GetEntryMessage(val keyAlias: String)

@Serializable
data class GetEntryResult(val key: String, val attributes: List<Pair<String, String>>)

@Serializable
data class ImportKeyMessage(val wrappedKeyAlias: String, val wrappedKey: String, val transformation: String, val spec: AlgorithmSpecMessage)

@Serializable
data class ImportKeyResult(val success: Boolean)

@Serializable
data class ListKeysResult(val keyAliases: List<String>)

@Serializable
data class EncryptDataMessage(val keyAlias: String, val transformation: String, val data: String)

@Serializable
data class EncryptDataResult(val data: String)

@Serializable
data class SignDataMessage(val keyAlias: String, val transformation: String, val data: String)

@Serializable
data class SignDataResult(val signature: String)

@Serializable
data class VerifySignatureMessage(val keyAlias: String, val transformation: String, val data: String, val signature: String)

@Serializable
data class VerifySignatureResult(val verified: Boolean)
